src/notification.rs

```diff
@@ -1,218 +1,163 @@
-    const DEPRECATED_TEXT: &str = "do not use deprecated field";
-
-    let singledomain = template
-        .render_to_string(&MustacheData {
-            commit: None,
-            singledomain: Some(DEPRECATED_TEXT.to_owned()),
-            singlehost: None,
-            siteamount: 42,
-            domains: vec![],
-            hosts: vec![],
-            sites: vec![],
-        })?
-        .contains(DEPRECATED_TEXT);
-    if singledomain {
-        crate::logger::warn(
-            "Replace singledomain with singlehost in template. singledomain will be removed in the future.",
-        );
+fn generate_commit_part(commit: Option<String>, template: Option<String>) -> Option<String> {
+    if let Some(template) = template {
+        commit.map(|commit| {
+            if template.contains("{commit}") {
+                template.replace("{commit}", &commit)
+            } else {
+                format!("{template}{commit}")
+            }
+        })
+    } else {
+        commit
     }
+}
 
-    let domains = template
-        .render_to_string(&MustacheData {
-            commit: None,
-            singledomain: None,
-            singlehost: None,
-            siteamount: 42,
-            domains: vec![DEPRECATED_TEXT.to_owned()],
-            hosts: vec![],
-            sites: vec![],
-        })?
-        .contains(DEPRECATED_TEXT);
-    if domains {
-        crate::logger::warn(
-            "Replace domains with hosts in template. domains will be removed in the future.",
-        );
-    }
+pub fn generate_text(
+    commit: Option<String>,
+    commit_template: Option<String>,
+    changed: Vec<Url>,
+) -> String {
+    let change_lines = generate_change_lines(changed);
 
-    let template = Some(template);
-    let any_empty = [
-        MustacheData::example_single(Some("666")).apply_to_template(template)?,
-        MustacheData::example_single(None).apply_to_template(template)?,
-        MustacheData::example_different(Some("666")).apply_to_template(template)?,
-        MustacheData::example_different(None).apply_to_template(template)?,
-        MustacheData::example_same(Some("666")).apply_to_template(template)?,
-        MustacheData::example_same(None).apply_to_template(template)?,
-    ]
-    .iter()
-    .any(std::string::String::is_empty);
-
-    if any_empty {
-        Err(anyhow::anyhow!("template produced empty notification text"))
+    if let Some(part) = generate_commit_part(commit, commit_template) {
+        format!("{part}\n\n{change_lines}")
     } else {
-        Ok(())
+        change_lines
     }
 }
 
 #[test]
-fn simple_template_is_valid() {
-    let template = mustache::compile_str("Hello {{name}}").unwrap();
-    validate_template(&template).unwrap();
+fn e2e() {
+    let result = generate_text(
+        Some("1234abc".to_owned()),
+        None,
+        vec![Url::parse("https://edjopato.de/").unwrap()],
+    );
+    assert_eq!(result, "1234abc\n\n- https://edjopato.de/");
 }
```

Comment:
```suggestion
fn e2e() {
    let result = generate_text(
        Some("1234abc".to_owned()),
        None,
        vec![Url::parse("https://edjopato.de/").unwrap()],
    );
    assert_eq!(result, "1234abc\n\n- https://edjopato.de/");
}
fn e2eTwoURL() {
    let result = generate_text(
        Some("1234abc".to_owned()),
        None,
        vec![Url::parse("https://edjopato.de/").unwrap(), Url::parse("https://teufelchen.eu/").unwrap()],
    );
    assert_eq!(result, "1234abc\n\n- https://edjopato.de/");
}
```

Even with this addition, still missing the behavior for an empty URL vector.

Reply:
e2e only checks that the parts are assembled together as expected. The parts are tested on their own (your added test would be [this one](https://github.com/EdJoPaTo/website-stalker/blob/78ab97f3a6cb9d6a1b05b4554618f2f24cb59f09/src/notification.rs#L107)).

A notification is only generated when there is at least one URL. Testing an empty input would test something that should not happen as it would notify that nothing happened.